algorithmic_efficiency/logging_utils.py

```diff
@@ -0,0 +1,492 @@
+                     accumulated_submission_time: float, goal_reached: bool,
+                     is_time_remaining: bool, training_complete: bool,
+                     early_stop: bool):
+    self._write_trial_metadata_file(workload, hyperparameters, trial_idx,
+                                    global_step, batch_size, latest_eval_result,
+                                    global_start_time,
+                                    accumulated_submission_time, goal_reached,
+                                    is_time_remaining, training_complete,
+                                    early_stop)
+
+  def workload_complete(self, score: float):
+    """At the end of the workload write COMPLETE to the metadata file."""
+    self.status = 'COMPLETE'
+    self._write_workload_metadata_file(score)
+
+  def _get_eval_measurements(self, workload: spec.Workload,
+                             hyperparameters: Optional[spec.Hyperparamters],
+                             trial_idx: int, global_step: int, batch_size: int,
+                             latest_eval_result: dict, global_start_time: float,
+                             accumulated_submission_time: float,
+                             goal_reached: bool, is_time_remaining: bool,
+                             training_complete: bool, early_stop: bool) -> dict:
+    """Collect all evaluation measurements and metadata in one dict."""
+    measurements = {}
+    measurements['workload'] = self.workload_name
+    measurements['trial_idx'] = trial_idx
+
+    # Record training measurements
+    measurements['datetime'] = datetime.now().isoformat()
+    measurements['accumulated_submission_time'] = accumulated_submission_time
+    if latest_eval_result:
+      for key, value in latest_eval_result.items():
+        measurements[key] = value
+    measurements['global_step'] = global_step
+    steps_per_epoch = workload.num_train_examples // batch_size
+    measurements['epoch'] = global_step / steps_per_epoch
+    measurements['steps_per_epoch'] = steps_per_epoch
+    measurements['global_start_time'] = global_start_time
+    measurements['goal_reached'] = goal_reached
+    measurements['is_time_remaining'] = is_time_remaining
+    measurements['early_stop'] = early_stop
+    measurements['training_complete'] = training_complete
+
+    # Record hyperparameters
+    measurements['batch_size'] = batch_size
+    if hyperparameters:
+      hparams_dict = hyperparameters._asdict()
+      # prefix every key with "hparam." to make more human-readable and to
+      # avoid overlap with other possible kys in the measurements dict.
+      hparams_dict = {f'hparam.{k}': v for k, v in hparams_dict.items()}
+      measurements.update(hparams_dict)
+
+    # Record workload properties
+    workload_properties = _get_workload_properties(workload)
+    measurements.update(workload_properties)
+
+    if FLAGS.extra_metadata:
+      extra_metadata = _get_extra_metadata_as_dict(FLAGS.extra_metadata)
```

Comment:
do we also need to write the same extra metadata to the measurements file on each row? is it sufficient to write it to  `metadata.json`?

Reply:
I'm thinking about this and a larger rework eliminate duplication. I'll reply with more details later.

Reply:
Yeah, this info gets duplicated. My thinking here is I'm trying to prioritize easy analysis.  The trade-off is that it saves users from needing to combine data from multiple files when doing analysis. I understand the other side of the trade-off is: information overload and no single source of truth. But I don't think those are severe concerns here. If you agree, then lets leave this in.

Reply:
I decided not to reduce duplication because of my goal to prioritize ease of analysis, but I'm still open to changing this. See the "[Summary of Measurements Collected](https://github.com/mlcommons/algorithmic-efficiency/blob/37521bf9b6f5a6a1a80dc59db38237b6a8745401/baselines/logging_example/README.md#summary-of-measurements-collected)" table to better understand the current duplicating of data.